var/spack/repos/builtin/packages/faiss/package.py

```diff
@@ -0,0 +1,177 @@
+    depends_on('py-setuptools', when='+python', type=('build', 'run'))
+    depends_on('swig',          when='+python', type='build')
+    depends_on('py-scipy',      when='+tests',  type=('build', 'run'))
+
+    # --- patch for v1.5.3 -----------------------------------------------
+    # faiss assumes that the "source directory" will always
+    # be called "faiss" (not spack-src or faiss-1.5.3)
+    # so, we will have to create a symlink to self (faiss did that in 1.6.3)
+    # and add an include path
+    patch('fixes-in-v1.5.3.patch', when='@1.5.3')
+
+    # --- patch for v1.6.3 -----------------------------------------------
+    # for v1.6.3, GPU build has a bug (two files need to be deleted)
+    # https://github.com/facebookresearch/faiss/issues/1159
+    # also, some include paths in gpu/tests/Makefile are missing
+    patch('fixes-in-v1.6.3.patch', when='@1.6.3')
+
+    # --------------------------------------------------------------------------
+    phases = ['configure', 'build', 'install']
+
+    def configure_args(self):
+
+        #TODO: ask spack team about the correct way to force this
+        if '+tests' in self.spec and '~python' in self.spec:
+            raise InstallError('Incorrect variants: +tests must be accompanied by +python')
+
+        args = []
+        if '+cuda' in self.spec:
+            args.append('--with-cuda={}'.format(self.spec['cuda'].prefix))
+        else:
+            args.append('--without-cuda')
+        return args
+
+    # --------------------------------------------------------------------------
+    def build(self, spec, prefix):
+
+        make()
+        if '+python' in self.spec:
+            make('-C', 'python')
+
+        # CPU tests
+        if '+tests' in self.spec:
+            os.chdir(os.path.join(self.stage.source_path, 'tests'))
+            make('tests', parallel=False)
+
+        # GPU tests
+        if '+tests' in self.spec and '+cuda' in self.spec:
+            os.chdir(os.path.join(self.stage.source_path, 'gpu', 'test'))
+            make('build', parallel=False)   # this target is added by the patch
+            make('demo_ivfpq_indexing_gpu', parallel=False)
+
+        os.chdir(self.stage.source_path)
+
+    # --------------------------------------------------------------------------
+    def install(self, spec, prefix):
+
+        make('install')
+
+        if '+python' in self.spec:
+
+            # faiss's suggested installation (using makefile) puts the
+            # python bindings in python prefix
+            # but, instead, we want to keep these files in the faiss prefix
+            # TODO: replace with make install once the PR is accepted
+            # https://github.com/facebookresearch/faiss/pull/1271
+            cmd = '{} setup.py install --prefix={}'
+            os.chdir(os.path.join(self.stage.source_path, 'python'))
+            os.system(cmd.format(self.spec['python'].command, self.prefix))
```

Comment:
```suggestion
            setup_py('install', '--prefix=' + prefix, '--single-version-externally-managed', '--root=/')
```
The extra args here will take care of the egg problem you noticed.

Reply:
Additional parameters address the egg problem, but somehow prevent installing executables (python entry functions i assume). not sure what's happening there.

Reply:
see for example the output of the python installation (`setup_py`) with and without your suggested additions.

- using `setup_py('install', '--prefix=' + prefix)`
```
==> [2020-06-30-11:26:14.556640, 91287] '/usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/python-3.7.6-35aupiwijpqsl7kmu76arlhijsltclx5/bin/python3.7' 'setup.py' '--no-user-cfg' 'install' '--prefix=/usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2'
running install
running bdist_egg
running egg_info
writing faiss.egg-info/PKG-INFO
writing dependency_links to faiss.egg-info/dependency_links.txt
writing requirements to faiss.egg-info/requires.txt
writing top-level names to faiss.egg-info/top_level.txt
reading manifest file 'faiss.egg-info/SOURCES.txt'
writing manifest file 'faiss.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-ppc64le/egg
running install_lib
running build_py
copying faiss/__init__.py -> build/lib/faiss
copying faiss/swigfaiss.py -> build/lib/faiss
copying faiss/_swigfaiss.so -> build/lib/faiss
creating build/bdist.linux-ppc64le/egg
creating build/bdist.linux-ppc64le/egg/faiss
copying build/lib/faiss/__init__.py -> build/bdist.linux-ppc64le/egg/faiss
copying build/lib/faiss/swigfaiss.py -> build/bdist.linux-ppc64le/egg/faiss
copying build/lib/faiss/_swigfaiss.so -> build/bdist.linux-ppc64le/egg/faiss
byte-compiling build/bdist.linux-ppc64le/egg/faiss/__init__.py to __init__.cpython-37.pyc
byte-compiling build/bdist.linux-ppc64le/egg/faiss/swigfaiss.py to swigfaiss.cpython-37.pyc
creating build/bdist.linux-ppc64le/egg/EGG-INFO
copying faiss.egg-info/PKG-INFO -> build/bdist.linux-ppc64le/egg/EGG-INFO
copying faiss.egg-info/SOURCES.txt -> build/bdist.linux-ppc64le/egg/EGG-INFO
copying faiss.egg-info/dependency_links.txt -> build/bdist.linux-ppc64le/egg/EGG-INFO
copying faiss.egg-info/requires.txt -> build/bdist.linux-ppc64le/egg/EGG-INFO
copying faiss.egg-info/top_level.txt -> build/bdist.linux-ppc64le/egg/EGG-INFO
writing build/bdist.linux-ppc64le/egg/EGG-INFO/native_libs.txt
zip_safe flag not set; analyzing archive contents...
creating 'dist/faiss-1.5.3-py3.7.egg' and adding 'build/bdist.linux-ppc64le/egg' to it
removing 'build/bdist.linux-ppc64le/egg' (and everything under it)
Creating /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages/site.py
Processing faiss-1.5.3-py3.7.egg
Copying faiss-1.5.3-py3.7.egg to /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages
Adding faiss 1.5.3 to easy-install.pth file

Installed /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages/faiss-1.5.3-py3.7.egg
Processing dependencies for faiss==1.5.3
Searching for numpy==1.19.0
Best match: numpy 1.19.0
Adding numpy 1.19.0 to easy-install.pth file
Installing f2py script to /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/bin
Installing f2py3 script to /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/bin
Installing f2py3.7 script to /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/bin

Using /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/python-3.7.6-35aupiwijpqsl7kmu76arlhijsltclx5/lib/python3.7/site-packages
Finished processing dependencies for faiss==1.5.3
==> [2020-06-30-11:26:16.528887, 91287] Installing faiss_tests to /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/bin
```

- using `setup_py('install', '--prefix=' + prefix, '--single-version-externally-managed', '--root=/')`
```
==> [2020-06-30-11:28:11.311685, 96252] '/usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/python-3.7.6-35aupiwijpqsl7kmu76arlhijsltclx5/bin/python3.7' 'setup.py' '--no-user-cfg' 'install' '--prefix=/usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2' '--single-version-externally-managed' '--root=/'
running install
running build
running build_py
copying faiss/__init__.py -> build/lib/faiss
copying faiss/swigfaiss.py -> build/lib/faiss
copying faiss/_swigfaiss.so -> build/lib/faiss
running install_lib
creating /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages/faiss
copying build/lib/faiss/__init__.py -> /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages/faiss
copying build/lib/faiss/swigfaiss.py -> /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages/faiss
copying build/lib/faiss/_swigfaiss.so -> /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages/faiss
byte-compiling /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages/faiss/__init__.py to __init__.cpython-37.pyc
byte-compiling /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages/faiss/swigfaiss.py to swigfaiss.cpython-37.pyc
running install_egg_info
running egg_info
writing faiss.egg-info/PKG-INFO
writing dependency_links to faiss.egg-info/dependency_links.txt
writing requirements to faiss.egg-info/requires.txt
writing top-level names to faiss.egg-info/top_level.txt
reading manifest file 'faiss.egg-info/SOURCES.txt'
writing manifest file 'faiss.egg-info/SOURCES.txt'
Copying faiss.egg-info to /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/lib/python3.7/site-packages/faiss-1.5.3-py3.7.egg-info
running install_scripts
==> [2020-06-30-11:28:12.087839, 96252] Installing faiss_tests to /usr/gapps/mummi/spack/opt/spack/linux-rhel7-power9le/gcc-7.3.1/faiss-1.5.3-u325zacmkjnselnnc7pwhn27jilczew2/bin
```

Reply:
@adamjstewart , trying to get your attention onto this.. :-)

Reply:
Sorry, this fell off my radar. I'm not sure why that flag would prevent the executables from being installed.